hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/sink/prometheus/PushGatewaySink.java

```diff
@@ -0,0 +1,191 @@
+
+  private static final Pattern SPLIT_PATTERN =
+      Pattern.compile("(?<!(^|[A-Z_]))(?=[A-Z])|(?<!^)(?=[A-Z][a-z])");
+
+  private static final String NUM_OPEN_CONNECTION_SPERUSER = "numopenconnectionsperuser";
+
+  private static final String NULL = "null";
+
+  private Map<String, String> groupingKey;
+  private PushGateway pg = null;
+  private String jobName;
+
+  @Override
+  public void init(SubsetConfiguration conf) {
+    // Get PushGateWay host configurations.
+    jobName = conf.getString(JOB_NAME, "hadoop-job");
+    final String serverHost = conf.getString(HOST_KEY);
+    final int serverPort = Integer.parseInt(conf.getString(PORT_KEY));
+
+    if (serverHost == null || serverHost.isEmpty() || serverPort < 1) {
+      throw new MetricsException(
+          "Invalid host/port configuration. Host: " + serverHost + " Port: " + serverPort);
+    }
+
+    groupingKey = parseGroupingKey(conf.getString(GROUP_KEY, ""));
+    pg = new PushGateway(serverHost + ':' + serverPort);
+  }
+
+  @Override
+  public void putMetrics(MetricsRecord metricsRecord) {
+    try {
+      CollectorRegistry registry = new CollectorRegistry();
+      for (AbstractMetric metrics : metricsRecord.metrics()) {
+        if (metrics.type() == COUNTER
+            || metrics.type() == GAUGE) {
+
+          String key = getMetricsName(
+              metricsRecord.name(), metrics.name()).replace(" ", "");
+
+          int tagSize = metricsRecord.tags().size();
+          String[] labelNames = new String[tagSize];
+          String[] labelValues = new String[tagSize];
+          int index = 0;
+          for (MetricsTag tag : metricsRecord.tags()) {
+            String tagName = tag.name().toLowerCase();
+
+            //ignore specific tag which includes sub-hierarchy
+            if (NUM_OPEN_CONNECTION_SPERUSER.equals(tagName)) {
+              continue;
+            }
+            labelNames[index] = tagName;
+            labelValues[index] =
+                tag.value() == null ? NULL : tag.value();
+            index++;
+          }
+
+          switch (metrics.type()) {
+          case GAUGE:
+            Gauge.build(key, key)
+                .labelNames(labelNames)
+                .register(registry)
+                .labels(labelValues)
+                .set(metrics.value().doubleValue());
+            break;
+          case COUNTER:
+            Counter.build(key, key)
+                .labelNames(labelNames)
+                .register(registry)
+                .labels(labelValues)
+                .inc(metrics.value().doubleValue());
+            break;
+          default:
+            break;
+          }
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("register succeed, metrics name is: {} Type is :{} tag is: {} tagValue is {} Value is : {}"
+                , key, metrics.type().toString(), Arrays.toString(labelNames)
+                , Arrays.toString(labelValues), metrics.value().toString());
+          }
```

Comment:
toString() can be removed here.
* metrics.type().toString() -> metrics.type()
* metrics.value().toString() -> metrics.value()

Reply:
> toString() can be removed here.
> 
> * metrics.type().toString() -> metrics.type()
> * metrics.value().toString() -> metrics.value()

well, I have changed it.